FUSETOOLS2-1045 - tutorial time

bfitzpat opened the pull request:
Exploring tracking of how long it would take to complete a particular tutorial. https://issues.redhat.com/browse/FUSETOOLS2-1045

Solution explorer used time tracking at the tutorial level (detailing the total time). The total is then shown here at the top-level page serving as a dashboard for all tutorial access and progress tracking.
![image](https://user-images.githubusercontent.com/530878/112878674-73a25280-9085-11eb-98b1-b71cd96e0cce.png)

We can do something similar by adding simple attributes (i.e. `{time=5}`) to various elements in the tutorial and then sum them up in the tutorial view in the description put beside each registered tutorial.

![image](https://user-images.githubusercontent.com/530878/112878715-7dc45100-9085-11eb-94d8-302c7a460e7a.png)

Signed-off-by: Brian Fitzpatrick <bfitzpat@redhat.com>

bfitzpat commented:
Now if the tutorial has time-boxed headings (that are summed up in the top-level node for the tutorial), we break out any time-boxed elements as children beneath.
![image](https://user-images.githubusercontent.com/530878/112890818-8f612500-9094-11eb-8849-6866976b295f.png)

bfitzpat commented:
Added the ability to quickly move to one of the time-bounded steps denoted in the treeview 
![time-box-demo-30-MAR-2021](https://user-images.githubusercontent.com/530878/113050787-79219a80-9162-11eb-908b-875c02b1443f.gif)

Right now this only works in markdown.

bfitzpat commented:
@pwright am wondering if you can help with the adoc part of this equation

Markdown offers a simple `{time=5}` notation that introduces an attribute on the heading element (i.e. `## Heading 2 {time=5}` generates an H2 element with the attribute time. 

AsciiDoc doesn't seem to want to play nice. :) 

I was hoping I could do something similar to what we did with Solution Patterns. 

```
:sectnums:

[time=5]
[id='my-step']
== My Step
:task-context: creating-project

Follow these steps to create a new project...
```
I tried this...
```= A Heading

Some text.

[time=6]
== A Second Heading That Should Be Time-marked

Lorem ipsum dolor sit amet. 

[time=16]
== A Third Heading That Should Be Time-marked

Lorem ipsum dolor sit amet. 

== Done! 

All done!
``` 
But it just ignores the notation.

pwright commented on src/didactManager.ts:
```diff
@@ -114,6 +115,34 @@ export class DidactManager {
+			const html : string | undefined = didactManager.active()?.getCurrentHTML();
+			const label : string = node.label as string;
+			if (html && label) {
+				const tags : string[] = ['h1', 'h2', 'h1'];
```
why repeat h1?

pwright commented:
Something like [role="time:6"] will add a class to the H2, not sure if that is something you could use?
```
<div class="sect1 time:6">
<h2 id="_a_second_heading_that_should_be_time_marked">A Second Heading That Should Be Time-marked</h2>
```

but I'm not sure what's stopping you from using the same technique as markdown?
isn't the didact code parsing the html for a json snippet, setting the time value  and removing it from the view?

bfitzpat commented:
Unfortunately it seems that the html parser we're using doesn't quite handle this kind of approach - https://jsfiddle.net/rbd904yj/1/

bfitzpat commented:
Have things working with asciidoc now -- and also added a couple of tests to make sure we're getting expected results
![image](https://user-images.githubusercontent.com/530878/113186216-99aa2d00-9214-11eb-81bb-6251f0c5ab23.png)

bfitzpat commented:
Added a modified hellojs example as a time-boxed example for markdown. Will add additional details in the wiki when we publish the extension with the new capabilities.

bfitzpat commented:
I am thinking we should disable the standard HelloJS example and just use the modified time-boxed one instead, since it serves two purposes that way. Any thoughts? @apupier ? @pwright ? @lhein ?

bfitzpat commented:
The windows test failure is weird. It works fine on my Windows 10 box. All tests pass. 
```
  walk through all the sendNamedTerminalAString commands in the demo
d:\Users\fitz\AppData\Local\Programs\Microsoft VS Code\resources\app\out\bootstrap-fork.js:5
validateTerminalResponse terminal newTerm executing text echo Hello Didact!
d:\Users\fitz\AppData\Local\Programs\Microsoft VS Code\resources\app\out\bootstrap-fork.js:5
2
-current terminal = newTerm
d:\Users\fitz\AppData\Local\Programs\Microsoft VS Code\resources\app\out\bootstrap-fork.js:5
      √ test terminal command "didact://?commandId=vscode.didact.sendNamedTerminalAString&text=newTerm$$echo%20Hello%20Didact!&completion=Opened%20a%20new%20terminal.": 1680ms
d:\Users\fitz\AppData\Local\Programs\Microsoft VS Code\resources\app\out\bootstrap-fork.js:5
validateTerminalResponse terminal terminal-name executing text terminal-command-to-execute
d:\Users\fitz\AppData\Local\Programs\Microsoft VS Code\resources\app\out\bootstrap-fork.js:5
3
-current terminal = terminal-name
d:\Users\fitz\AppData\Local\Programs\Microsoft VS Code\resources\app\out\bootstrap-fork.js:5
      √ test terminal command "didact://?commandId=vscode.didact.sendNamedTerminalAString&user=terminal-name$$terminal-command-to-execute&completion=Opened%20a%20new%20terminal.": 4626ms
d:\Users\fitz\AppData\Local\Programs\Microsoft VS Code\resources\app\out\bootstrap-fork.js:5
  Suite duration: 6.309 s, Tests: 2
d:\Users\fitz\AppData\Local\Programs\Microsoft VS Code\resources\app\out\bootstrap-fork.js:5
```

bfitzpat commented:
Looks like I have stabilized the tests more.

apupier reviewed (changes requested):
- Why the helloJS tutorial has been mostly duplicated, just adding the time? Why not simply updating the existig one with time information?
- when there are several headings with same name, the tree displayed a single of them:
![Screenshot from 2021-04-02 11-55-54](https://user-images.githubusercontent.com/1105127/113405611-67323880-93aa-11eb-9f83-8af8feeed646.png)
- optional: when a time attribute has non-numeric value, NaN is displayed in tree. Wondering if it is better to not display it at all. On the other hands it is makign it more visible to have it

bfitzpat commented on src/nodeProvider.ts:
```diff
@@ -38,21 +40,23 @@ export class DidactNodeProvider implements vscode.TreeDataProvider<TreeNode> {
-        } else {
-            // assume it's a category
-            if (element.label && !(element instanceof TutorialNode)) {
+	public async getChildren(element?: any): Promise<any[]> {
```
because we don't just return TreeNodes - they're HeadingNodes or TutorialNodes. TutorialNode is a TreeNode, but HeadingNode is actually a vscode.TreeItem. it's a mix of things

bfitzpat replied:
Will try and clean things up with https://issues.redhat.com/browse/FUSETOOLS2-1050

apupier replied:
if there are two types, TypeScript is providign th epossibility to provide both of them with `HeadingNodes | TutorialNodes`

it sounds better to me to handle it inside this PR as it is introducing technical debt that sounds ot be easuily avoided. But fine to handle it in another task as longas it is done shortly after.

bfitzpat replied:
@apupier it's more than that, so I will handle it in a separate follow-on PR

bfitzpat commented on src/extensionFunctions.ts:
```diff
@@ -974,3 +988,75 @@ export async function pasteClipboardToNewTextFile() : Promise<void> {
+	return -1;
+}
+
+export async function getHeadingsForDidactFileUri(uri: vscode.Uri) : Promise<any[] | undefined> {
+	if (uri) {
+		let content : string | undefined | void = undefined;
+		if (uri.scheme === 'file') {
+			content = await loadFileWithRetry(uri);
+		} else if (uri.scheme === 'http' || uri.scheme === 'https'){
+			content = await loadFileFromHTTPWithRetry(uri);
+		}
+		if (content && !isAsciiDoc(uri.toString())) {
+			return getTimeElementsForMD(content);
+		} else if (content && isAsciiDoc(uri.toString())) {
+			return getTimeElementsForAdoc(content);
+		}
+	}
+	return undefined;
+}
```
In this case, you are correct. In the tree we only show the time elements that resolve to H1-H4 HTML heading elements so I conflated the two areas a bit. This method returns ALL time elements from the didact files that we know about. We don't resolve them specifically to headings at this point. Will change the name of the method to `getTimeElementsForDidactFileUri`.

bfitzpat commented on src/extension.ts:
```diff
@@ -42,7 +42,7 @@ export async function activate(context: vscode.ExtensionContext): Promise<void>
 	context.subscriptions.push(vscode.commands.registerCommand(extensionFunctions.CREATE_WORKSPACE_FOLDER_COMMAND, extensionFunctions.createTemporaryFolderAsWorkspaceRoot));
 	context.subscriptions.push(vscode.commands.registerCommand(extensionFunctions.OPEN_TUTORIAL_COMMAND, extensionFunctions.openDidactWithDefault));
-	context.subscriptions.push(vscode.commands.registerCommand(extensionFunctions.START_DIDACT_COMMAND, extensionFunctions.startDidact));
+	context.subscriptions.push(vscode.commands.registerCommand(extensionFunctions.START_DIDACT_COMMAND, extensionFunctions.findOrCreateDidact));
```
It does the expected behavior from the user's perspective. If the didact file is already open, we don't want to open it again. If it's not open, we want to open it. Either way, it "starts" the didact file without completely changing the way that users currently think of "starting" a didact file.

apupier replied:
When I read the `findOrCreateDidact`, I do not understand the behavior that you explained.
I understand "create Didact"  as create a Didact file, a *.didact.adoc or *.didact.md; not create a rendering of the Didact file inside a webview.

bfitzpat replied:
> i think that it needs to be documented as a limitation or implemented

@apupier I'm curious where you think it should be documented -- but even more interested in how you think it might be implemented. At this point, it is not possible to automatically have an id applied to every heading.

bfitzpat replied:
> When I read the `findOrCreateDidact`, I do not understand the behavior that you explained.

@apupier what would you prefer? `findOrStartDidact`?

apupier replied:
based on implementation:
```
	if (uri) {
		const parentPanel = didactManager.getByUri(uri);
		if (parentPanel) {
			parentPanel._panel?.reveal();
		} else {
			await startDidact(uri, viewColumn);
		}
	}
```
Maybe `revealOrStartDidact` , `revealOrStartDidactTutorial`

bfitzpat replied:
went with `revealOrStartDidactByURI`

bfitzpat commented on src/nodeProvider.ts:
```diff
@@ -38,21 +40,23 @@ export class DidactNodeProvider implements vscode.TreeDataProvider<TreeNode> {
 
 	// add a child to the list of nodes
-	public addChild(newNode: TreeNode, disableRefresh = false, oldNodes: TreeNode[] = this.treeNodes ): TreeNode[] {
+	public addChild(newNode: any, disableRefresh = false, oldNodes: any[] = this.treeNodes ): any[] {
```
>     * when there are several headings with same name, the tree displayed a single of them:

At this time, that's due to the fact that headings do not get unique IDs so we are forced to do a text search for them. We search for the first one since there's no way to know which one they may have meant.

bfitzpat replied:
As for the Any and losing type safety, there are some problems with the current implementation. I would like to address those in a follow-up PR to rearrange the tree class structure a bit to simplify things. Right now there is a mix of vscode.TreeItem and TreeNode implementations as children, which makes this a bit clunky. Let's move it to https://issues.redhat.com/browse/FUSETOOLS2-1050

bfitzpat commented:
>     * Why the helloJS tutorial has been mostly duplicated, just adding the time? Why not simply updating the existig one with time information?

To make the effort more explicit. To the user it won't change anything.

bfitzpat commented:
> optional: when a time attribute has non-numeric value, NaN is displayed in tree. Wondering if it is better to not display it at all. On the other hands it is makign it more visible to have it

Good idea. I will do two things. One, I won't include the NaN in the tree. Two, I will include a warning message in the Didact output channel explaining which heading had the issue.

apupier commented:
> > Why the helloJS tutorial has been mostly duplicated, just adding the time? Why not simply updating the existig one with time information?
> 
> To the user it's not going to matter. Initially I felt like I wanted to spin off a new version of the example. Ultimately it's not going to change anything from the user's perspective beyond showing the time elements. I don't see the point in changing it back at this point.

it seems interesting that the users benefit from the feature.

it also avoids to have duplicated code and will ease maintenance.

apupier commented:
> > ```
> > * when there are several headings with same name, the tree displayed a single of them:
> > ```
> 
> 
> At this time, that's due to the fact that headings do not get unique IDs so we are forced to do a text search for them. We search for the first one since there's no way to know which one they may have meant.

i think that it needs to be documented as a limitation or implemented.

apupier reviewed (commented):
provided comments

bfitzpat commented:
> > > Why the helloJS tutorial has been mostly duplicated, just adding the time? Why not simply updating the existig one with time information?
> it also avoids to have duplicated code and will ease maintenance.

Done. Moved change to existing tutorial. Removed duplicate

apupier commented:
> > > > ```
> > > > * when there are several headings with same name, the tree displayed a single of them:
> > > > ```
> > > 
> > > 
> > > At this time, that's due to the fact that headings do not get unique IDs so we are forced to do a text search for them. We search for the first one since there's no way to know which one they may have meant.
> > 
> > 
> > i think that it needs to be documented as a limitation or implemented.
> >
> @apupier I'm curious where you think it should be documented -- but even more interested in how you think it might be implemented. At this point, it is not possible to automatically have an id applied to every heading.

There is a use case which is not that much uncommon which is not working. I provided an [example](https://github.com/redhat-developer/vscode-didact/pull/490#pullrequestreview-626884775) which is something has fair probability to be used. Consequently, i think it is important that it is either documented that the use case is not covered or that it is implemented.

it can be documented in changelog, in readme, in wiki.

On the how i'm not a specialist of markdown/asciidoc and genertaed html but I can imagine several possibilities:
* during generation of html, provide a more specific id based on the "heading path"
* use the full path of the element in the dom and not just the id.

bfitzpat commented:
> There is a use case which is not that much uncommon which is not working. I provided an [example](https://github.com/redhat-developer/vscode-didact/pull/490#pullrequestreview-626884775) which is something has fair probability to be used. Consequently, i think it is important that it is either documented that the use case is not covered or that it is implemented.

Yes, I understand the use case. 

> it can be documented in changelog, in readme, in wiki.

Thanks. That's what I was after here. 
 
> On the how i'm not a specialist of markdown/asciidoc and genertaed html

The issue is with rendering markdown and asciidoc as HTML. The different renderers work differently. This is not something we can resolve in this issue, which has already become overgrown. 

I have spun off https://issues.redhat.com/browse/FUSETOOLS2-1053 with some of the issues encountered with the markdown renderer for further discussion.

bfitzpat commented:
Can't really edit a wiki in a PR easily, but I created https://issues.redhat.com/browse/FUSETOOLS2-1054 to make sure that we document how to use this new feature as well as its current limitations
